### PR TITLE
Typed py geth

### LINKED ISSUE #116: Add types
* `py-geth` Version: 3.10
* `go-ethereum` Version: 1.10.25-stable
* Python Version: 3.9.13
* OS: macOS

### What was wrong?

Currently, when using `py-geth` as a library, we have to put `type: ignore` statements on all of the imports:

https://github.com/ApeWorX/ape/blob/main/src/ape_geth/provider.py#L16-L20

Adding types will help us verify we are using the library correctly as well as assist us it looks better (low priority).

#### Cute Animal Picture

chicken

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?

`py-geth` had no typing. 

Closes #116 

### How was it fixed?

Added typing

Notable updates:
 - Functions in `accounts.py` 1-2 args + the `geth_kwargs` dict, and returned `bytes`. They've been updated to accept only `kwargs` and return strings.
 - `get_geth_version_info_string` now returns `str` instead of `bytes`
 - Removed previously-deprecated items - `LiveGethProcess`, geth kwargs `ipc_api`, `miner_threads` and the `--ssh` flag.
 - `write_genesis_file` now takes `genesis_data` as a `dict` or `kwargs`
 - Genesis `gasLimit` updated to geth's default `0x47e7c4`
 - Have all raised exceptions inherit from a base `PyGethException`

See release notes for full detail.

Thanks, @DavidRomanovizc for your assistance in the initial test PR #161

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/py-geth/blob/main/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

**Review by fselmo — COMMENTED** [review]
lgtm... I think there's value in adding a test similar to [this one in web3.py](https://github.com/ethereum/web3.py/blob/main/tests/core/test_library_files.py) that checks whether we are raising any builtin Python exceptions. This helps future-proof code simple it's really easy to write code that raises `ValueError` and when reviewing it can make a lot of sense until we realize we should use `PyGethValueError` instead.

File: geth/utils/networking.py [hunk 1]
```diff
@@ -57,4 +61,4 @@ def wait_for_http_connection(port: int, timeout: int = 5) -> None:
             else:
                 break
         else:
-            raise ValueError("Unable to establish HTTP connection")
```
Comment (fselmo) [inline geth/utils/networking.py:55]: 👌🏼

File: tests/core/test_library_files.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import os
+
+PY_GETH_PATH = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)), "..", "..", "geth"
+)
+DEFAULT_EXCEPTIONS = (
```
Comment (fselmo) [inline tests/core/test_library_files.py:6]: We should probably add all the ones we've already created in here:

- `AttributeError`
- `KeyError`
- `TypeError`
- `ValueError` 
- `OSError`
- `NotImplementedError`
- `FileNotFoundError`

**Review by fselmo — APPROVED** [review]
lgtm... we should add all the Python builtin exceptions we've already extended into internal py-geth exceptions to that default list in the test

File: tests/core/test_library_files.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    TypeError,
+    ValueError,
+)
+
+
+def test_no_default_exceptions_are_raised_within_py_geth():
+    for root, _dirs, files in os.walk(PY_GETH_PATH):
+        for file in files:
+            if file.endswith(".py"):
+                file_path = os.path.join(root, file)
```
Comment (kclowes) [inline tests/core/test_library_files.py:27]: nit: I don't think the file path is quite right: 
image

**Review by kclowes — COMMENTED** [review]
Nice work! I _believe_ we use py-geth to generate our fixtures in web3. Did you check that to make sure any ripple effects are handled? Otherwise, I left some small inline comments

**Comment by kclowes** [comment]
It looks like the README needs to be updated as well since `LiveGethProcess` doesn't exist any more

**Comment by kclowes** [comment]
This may be an existing bug, or just something weird in my env, but running through the README examples, when I ask the `DevGethProcess` for accounts, I get a double of the same address:
```
>>> from geth import DevGethProcess
>>> g = DevGethProcess()
>>> g.accounts
('0xa95c66bff0c92edecb365c23d86f687026b48d79',
 '0xa95c66bff0c92edecb365c23d86f687026b48d79')
```

File: geth/utils/encoding.py [hunk 4]
```diff
@@ -1,41 +1,48 @@
...
+        encoded = codecs.encode(value, encoding)
+        if isinstance(encoded, (bytes, bytearray)):
+            return encoded
+        else:
+            raise PyGethTypeError(
+                f"Encoding {encoding!r} produced non-binary result: {encoded!r}"
+            )
     else:
-        raise TypeError(f"Unsupported type: {type(value)}")
+        raise PyGethTypeError(f"Unsupported type: {type(value)}")
```
Comment (reedsa) [inline geth/utils/encoding.py:33]: For further clarity, could add the expected type(s) to the message.

File: geth/utils/networking.py [hunk 5]
```diff
@@ -52,4 +61,4 @@ def wait_for_http_connection(port, timeout=5):
...
+            raise PyGethValueError("Unable to establish HTTP connection")
```
Comment (reedsa) [inline geth/utils/networking.py:64]:
```suggestion
            raise PyGethValueError(f"Unable to establish HTTP connection, timed out after {timeout} seconds.")
```

File: geth/utils/encoding.py [hunk 6]
```diff
@@ -1,41 +1,48 @@
...
-        return value
-    elif is_binary(value):
+def force_text(value: bytes | bytearray | str, encoding: str = "iso-8859-1") -> str:
+    if isinstance(value, (bytes, bytearray)):
         return codecs.decode(value, encoding)
+    elif isinstance(value, str):
+        return value
     else:
-        raise TypeError(f"Unsupported type: {type(value)}")
+        raise PyGethTypeError(f"Unsupported type: {type(value)}")
```
Comment (reedsa) [inline geth/utils/encoding.py:42]: Could add expected types in the error message.
```suggestion
        raise PyGethTypeError(f"Unsupported type: {type(value)}, expected value to be bytes, bytearray or str")
```

**Review by reedsa — APPROVED** [review]
Very nice work here, just left a few optional suggestions!

File: geth/exceptions.py [hunk 7]
```diff
@@ -39,3 +77,40 @@ def __str__(self):
...
+class PyGethGethError(PyGethException, GethError):
+    def __init__(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ):
+        GethError.__init__(*args, **kwargs)
+
+
+class PyGethAttributeError(PyGethException, AttributeError):
```
[the remark this answers is not in the document]
Comment (pacrob) [reply geth/exceptions.py:91]: My understanding of this exception grouping is that we preserve the ability to catch whatever exceptions a user might want to catch (if they want to handle `AttributeError` differently than `AssertionError`, for example) but just add on the ability to catch all py-geth errors at once.

And just semantically, it seems like a `PyGethAttributeError` should be an instance of `AttributeError`, no?

File: geth/process.py [hunk 8]
```diff
@@ -1,27 +1,34 @@
...
-try:
-    from urllib.request import (
-        URLError,
-        urlopen,
-    )
-except ImportError:
-    from urllib2 import (
-        urlopen,
-        URLError,
-    )
```
[the remark this answers is not in the document]
Comment (pacrob) [reply geth/process.py:23]: urllib2 is python2 - https://docs.python.org/2/library/urllib2.html

File: tests/core/test_library_files.py (hunk 3 above)
Comment (pacrob) [reply tests/core/test_library_files.py:27]: Hrm. Not sure how you're getting that error - the error on line 42 of `accounts.py` is already a `PyGethValueError`. If I change it to just a `ValueError` and run the test again, it catches it and gives the error message you have.

**Comment by pacrob** [comment]
> It looks like the README needs to be updated as well since `LiveGethProcess` doesn't exist any more

I'm not seeing any reference to `LiveGethProcess` anywhere except changelog and newsfragments. Where are you seeing it?

**Comment by pacrob** [comment]
> This may be an existing bug, or just something weird in my env, but running through the README examples, when I ask the `DevGethProcess` for accounts, I get a double of the same address:
> 
> ```
> >>> from geth import DevGethProcess
> >>> g = DevGethProcess()
> >>> g.accounts
> ('0xa95c66bff0c92edecb365c23d86f687026b48d79',
>  '0xa95c66bff0c92edecb365c23d86f687026b48d79')
> ```



> This may be an existing bug, or just something weird in my env, but running through the README examples, when I ask the `DevGethProcess` for accounts, I get a double of the same address:
> 
> ```
> >>> from geth import DevGethProcess
> >>> g = DevGethProcess()
> >>> g.accounts
> ('0xa95c66bff0c92edecb365c23d86f687026b48d79',
>  '0xa95c66bff0c92edecb365c23d86f687026b48d79')
> ```

That is pre-existing - I get the same with the current release `4.4.0` Not sure why, but maybe related to the 2 instances of `password` on ln 157 of `accounts.py`?

File: geth/utils/encoding.py (hunk 6 above)
Comment (pacrob) [reply geth/utils/encoding.py]: added

File: geth/utils/networking.py (hunk 5 above)
Comment (pacrob) [reply geth/utils/networking.py]: added

File: geth/utils/encoding.py (hunk 4 above)
Comment (pacrob) [reply geth/utils/encoding.py]: added

File: geth/exceptions.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fselmo) [reply geth/exceptions.py:91]: > My understanding of this exception grouping is that we preserve the ability to catch whatever exceptions a user might want to catch (if they want to handle `AttributeError` differently than `AssertionError`, for example) but just add on the ability to catch all py-geth errors at once.
> 
> And just semantically, it seems like a `PyGethAttributeError` should be an instance of `AttributeError`, no?

+1... this is how we set it up in *web3.py* as well

File: geth/exceptions.py [hunk 9]
```diff
@@ -1,23 +1,61 @@
...
+    ):
+        super().__init__(*args)
 
+        # Assign properties of PyGethException
+        self.user_message = user_message
 
-class GethError(Exception):
-    message = DEFAULT_MESSAGE
+
+class GethError:
```
[the remark this answers is not in the document]
Comment (pacrob) [reply geth/exceptions.py]: added

File: geth/mixins.py [hunk 10]
```diff
@@ -74,58 +93,68 @@ class InterceptedStreamsMixin:
     stderr lines into some set of provided callback functions.
     """
 
-    stdout_callbacks = None
-    stderr_callbacks = None
+    stdout_callbacks: list[Callable[..., Any]]
+    stderr_callbacks: list[Callable[..., Any]]
```
[the remark this answers is not in the document]
Comment (pacrob) [reply geth/mixins.py]: tightened

**Comment by pacrob** [comment]
> Nice work! I _believe_ we use py-geth to generate our fixtures in web3. Did you check that to make sure any ripple effects are handled? Otherwise, I left some small inline comments

Test pr here: https://github.com/ethereum/web3.py/pull/3417

Passes all web3 tests with current `typed-py-geth` branch as dep.

**Review by kclowes — APPROVED** [review]
LGTM! Nice work!

**Pull request merged**